Issue #51 interpolator

jrgnicho opened the pull request:
This PR adds the CartesianInterpolator class which addresses the following high level objectives:
- Generates a dense trajectory by interpolating between the points of a dense trajectory in cartesian space.
- It assumes that the tool moves at a constant linear speed through the entire trajectory.
- It uses the user defined tool speed and time interval in order to determine the trajectory density (how close the points are to one another).
- No parabolic blending or accelerations profiles are being generated at the moment.
- This class should help guide designing a template for a future interpolator interfaces.

jrgnicho committed:
Added the CartesianInterpolator class

jrgnicho committed:
fixed minor documentation error

jrgnicho committed:
added unit test for the cartesian interpolator

jrgnicho committed:
added dependencies for cartesian interpolator unit test

jrgnicho committed:
completed unit test for cartesian interpolator

Jmeyer1292 commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+namespace descartes_trajectory
+{
+
+class TrajectorySegment
```
As this class and its children are not part of the public interface of the cartesian interpolater, would it make sense to put these in a separate header file?

jrgnicho replied:
True, they they are only ever used by the interpolator so they should be hidden from the interface.  Not sure that a separate header file is the way to go, they could be placed as internal classes of CartesianInterpolator that aren't accessible to the outside

Jmeyer1292 commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+   * @param interpolated_traj the interpolated trajectory
+   */
+  bool interpolate(const std::vector<descartes_trajectory::CartTrajectoryPt>& coarse_traj,
+                   std::vector<descartes_core::TrajectoryPtPtr>& interpolated_traj);
```
Just out of curiosity, why do we take input by value and return by shared pointer? As far as I can tell all this will ever do is return CartTrajPt's itself.

jrgnicho replied:
This should facilitate implementing a planning pipeline where the input is specified in CartesianPts.  Thus the workflow would be:
CartesianPts -> Interpolator -> TrajectoryPtPtrs -> Planner -> TrajectoryPtPtrs

Jmeyer1292 commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+  tf::Transform p;
+  Eigen::Affine3d eigen_pose;
+  tf::Quaternion q;
+  for(std::size_t i = 1; i < num_steps; i++)
```
Might be worth mentioning somewhere that this function appears to return the range:  (begin, end]

Jmeyer1292 commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+
+      if(i == segment_times.size() - 1)
+      {
+        interpolated_traj.push_back(TrajectoryPtPtr(new CartTrajectoryPt(ps)));
```
I'm a little confused on what's going on here. On the last iteration you add the start point of the last interpolation to the end? Should this be the variable "pe" instead?

jrgnicho replied:
Good catch, it's supposed to be "pe" .

Jmeyer1292 commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+
+  tf::Vector3 vel = (end_pose_.getOrigin() - start_pose_.getOrigin())/total_time_;
+  double time_step = interpolation_step_/total_time_;
```
This is kind of a confusing name for a variable given that you also have interpolation_step_. Perhaps something more specific like "slerp_step_ratio" or something like that.

jrgnicho replied:
I'm not so sure that the suggested name is that much more clarifying.

Jmeyer1292 commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+{
+public:
+  CartesianInterpolator();
+  virtual ~CartesianInterpolator();
```
Do you anticipate people using this class as a base class?

jrgnicho replied:
No, but it should help guide in defining the behavior needed in an interpolator interface.

Jmeyer1292 commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+                double interpolation_step);
+
+  virtual ~LinearSegment();
+  virtual bool interpolate(std::vector<descartes_core::TrajectoryPtPtr>& segment_points);
```
You'd have to change the interface, but this function is a great candidate for a const tag because it doesn't appear to mutate any of its state.

shaun-edwards replied:
+1

shaun-edwards commented:
I want to review this as well before it is accepted.  I should be able to do this tomorrow (Monday) night.

jrgnicho committed:
correction to addition of last point in the interpolated trajectory

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+
+};
+
+class CartesianInterpolator
```
Need class doxygen comment.

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+
+  }
+
+  virtual bool interpolate(std::vector<descartes_core::TrajectoryPtPtr>& segment_points) = 0;
```
What does this method do.  It looks like you intend this to be an interface.  If so, you need to document that interface.

jrgnicho replied:
I intended for this to be an interface for various segment (point to point) interpolation implementations. Initially I wanted to implement a Linear and Parabolic blend types but the Parabolic turned out to be more difficult than I had anticipated.  Agreed, this needs to be documented.

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+{
+public:
+
+  BlendSegment(const descartes_trajectory::CartTrajectoryPt& start,
```
This method is cartesian point specific.  We must be able to create segments that have joint/cartesian end-points.

jrgnicho replied:
+1

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+
+  BlendSegment(const descartes_trajectory::CartTrajectoryPt& start,
+               const descartes_trajectory::CartTrajectoryPt& end,
+               double duration,
```
Is `duration` required?  Doesn't the `TrajectoryPt` class already have this (it should)?

jrgnicho replied:
The TrajectoryPt has a transition pointer which only imposes motion constraints but no time value as far as I know.  Plus judging from previous discussions (see #10) these where going to be replaced eventually.

shaun-edwards replied:
The transitions should be replaced/removed.  This works makes them redundant.  The `TrajectoryPt` should capture the point velocity/timing at a minimum.  This would allow us to better prune our search tree.  I'm not sure how velocity generalizes to joint/cartesian points (maybe it doesn't have to).  Timing is pretty generic.

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+  BlendSegment(const descartes_trajectory::CartTrajectoryPt& start,
+               const descartes_trajectory::CartTrajectoryPt& end,
+               double duration,
+               double time_step);
```
Do we want a `time-step` or does something else make sense.  It could be a _distance_ or a number of interpolation points.

jrgnicho replied:
The math that I had found for generating parabolic blends defined equations that were heavily paramerized on discrete time steps.  It may be possible to go work our way from number of points into time step size, however I'm not so sure that in the case of distance it'd be so straight forward. 

For the case of Linear Segments, a case could be made for using distance over time-step since you could generate evenly spaced segments.

ClayFlannigan replied:
Outsider's perspective: we should implement a cubic spline such as [this](https://gist.github.com/layzerar/4132153) (minus the OpenCV dependency).  Cubic splines are the "standard" approach for smooth trajectories; they are twice differentiable meaning they have bounded accelerations.  The initial results I've seen from the blending work seemed to be velocity limited due to "non-smooth" trajectories.

shaun-edwards replied:
KDL has cubic interpolation.  We should use this.

shaun-edwards replied:
See issue #48 regarding the problems with blending.

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+
+protected:
+
+  tf::Transform start_pose_;
```
Should we use `tf` or `Eigen` internally.  `tf` may utilize Eigen, but our core planner stuff works off of `Eigen`.

jrgnicho replied:
+1 for using Eigen

shaun-edwards commented on descartes_trajectory/include/descartes_trajectory/cartesian_interpolator.h:
```diff
@@ -0,0 +1,121 @@
+   * @param coarse_traj the coarse trajectory that is to be interpolated
+   * @param interpolated_traj the interpolated trajectory
+   */
+  bool interpolate(const std::vector<descartes_trajectory::CartTrajectoryPt>& coarse_traj,
```
This method has two issues:
1. The input is only Cartesian points.  We should support hybrid paths
2. If you fix the problem above, there won't be any difference between the coarse and interpolated trajectories.  Does the sparse trajectory need it's own type?  Either a typedef or perhaps a container class?

jrgnicho replied:
I think it'd be best to define a new  type (subclass of TrajectoryPt) that contains sufficient information for how to go about the interpolation.  Plus if it had to be extended/modified  we would just make those changes to that class instead of TrajectoryPt which could break other implementations.

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+ * limitations under the License.
+ */
+
+#include <console_bridge/console.h>
```
I thought we decided not to use the `console_bridge`.  There's an issue somewhere.

jrgnicho replied:
+1.  It will be removed

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+
+namespace descartes_trajectory
+{
+LinearSegment::LinearSegment(const tf::Transform& start_pose,
```
Should we check for negative `total_time` and `interpolation_time`?

jrgnicho replied:
+1

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+
+  if(num_steps <= MIN_NUM_STEPS)
+  {
+    ROS_ERROR_STREAM("Interpolation steps value of "<<num_steps<<" is invalid");
```
You may want to print out `MIN_NUM_STEPS`, just in case it's never 1.

jrgnicho replied:
+1

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+
+  ROS_DEBUG_STREAM("Segment contains "<<segment_points.size()<<" interpolated points for "<<num_steps<<" time steps");
+
+  return true;
```
You could easily refactor this method to return from a single location just  by rearranging the `if` statement.

jrgnicho replied:
+1

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+  return true;
+}
+
+CartesianInterpolator::CartesianInterpolator()
```
Is this constructor hidden?  Using this constructor will result in undefined class variables (or rather random values).

jrgnicho replied:
It's not hidden

shaun-edwards commented on descartes_trajectory/src/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,199 @@
+  return true;
+}
+
+bool CartesianInterpolator::initialize(descartes_core::RobotModelConstPtr robot_model,
```
See note above about checking for negative values.

jrgnicho replied:
+1

shaun-edwards commented on descartes_trajectory/test/cartesian_interpolator.cpp:
```diff
@@ -0,0 +1,82 @@
+  EXPECT_TRUE(c.interpolate(coarse_traj,interpolated_traj));
+
+  // expected point count
+  EXPECT_TRUE(interpolated_traj.size() == expected_points);
```
This is a good check, but we could still be missing something.  I did something similar for point sampling and that caused some problems, if you remember.

shaun-edwards commented:
My general feeling is that this code is too specific to Cartesian trajectories.  Support for coarse/dense hybrid trajectories is important.

jrgnicho committed:
- removed console_bridge header
- Changed input coarse trajectory type to TrajectoryPtPtrs in order to support hydrid trajectories
- Segment classes now take eigen poses instead of tf::Transforms
- Checks for negative time values were added
- Cartesian interpolator constructor initializes members to arbitrary values

jrgnicho committed:
updated cartesian interpolator unit test to reflect the changes made to this class

Commit:
Changed Descartes objects to accept shared ptrs by value instead of by reference

Commit:
started implementation of  blend segments

Commit:
cartesia intepolator generates blending segments and in between linear segments

Commit:
moved TrajectorySegment classes into their own header file

jrgnicho commented:
The high level changes in the last set of commits are as follows:
- Implements blend regions with constant acceleration of the tool.
- Updated unit test for testing the cartesian interpolator
- Moved the TrajectorySegment classes into their own header and source file

@shaun-edwards  and @Jmeyer1292  please review

shaun-edwards commented:
@jrgnicho, we are going to back-burner this one for now.

Jmeyer1292 commented:
I think this kind of functionality belongs at a higher level than Descartes.
